processing/src/main/java/org/apache/druid/query/filter/EqualityFilter.java

```diff
@@ -412,7 +431,17 @@ private Supplier<DruidObjectPredicate<String>> makeStringPredicateSupplier()
         if (castForComparison == null) {
           return DruidObjectPredicate.alwaysFalseWithNullUnknown();
         }
-        return DruidObjectPredicate.equalTo(castForComparison.asString());
+        // when matching strings to numeric match values, use numeric comparator to implicitly cast the string to number
+        if (matchValue.type().isNumeric()) {
+          return value -> {
+            if (value == null) {
+              return DruidPredicateMatch.UNKNOWN;
+            }
+            return DruidPredicateMatch.of(StringComparators.NUMERIC.compare(value, castForComparison.asString()) == 0);
```

Comment:
`StringComparators.NUMERIC` is full of interesting calls - like: `convertStringToBigDecimal` - seems like something which could possibly generate some garbage collector pressure ...especially if it gets used inside a for-each-row comparator

Reply:
This change is trying to make this filter when comparing strings to numerics consistent with the behavior of the old `BoundFilter`, which was what was previously always used for comparing strings to numeric values until these new filters were added, equality was planned to something like `1.0 <= stringCol <= 1.0`. `BoundFilter` uses this comparator when comparing strings to numeric values, https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/filter/BoundFilter.java#L265. The goal here is to make the behavior the same regardless of how `sqlUseBoundAndSelectors` (the flag that controls whether or not we use the new or old filters is set).

Casting the string column to a double column is also relatively expensive, currently we use https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/math/expr/ExprEval.java#L627, where `Doubles.tryParse` involves a regex. This could probably be done more efficiently, but again goes back to requiring more dramatic changes be made in the native layer which I was trying to avoid for now.

Reply:
the `TypedInFilter` class 
* knows in which type the match should happen `matchValueType`
* the values it has should already be preapared to be used with that

> Casting the string column to a double column is also relatively expensive
with a virtual column that might be true;
but I think that casting will happen quite a lot with the current approach as well:
let me try to explain what I see right now:
* load an AVL tree
* do comparisions with a comparator
  * which will create a [BigDecimal](/blob/7ac0862287270fcb0cfb33b01a31ae2e650621ef/processing/src/main/java/org/apache/druid/query/ordering/StringComparators.java#L398-L399) when
     * not a comparision against `null`
     * the number(s) are not a `long`
  * that's kinda like all interesting cases...
  * due to the AVL tree lookup there will be `O(logN)` scaler
  * the `BigDecimal` creation will also happen in cases the comparator return `false`

I wonder why not:
* remove the AVL tree stuff
* the sorted values should be `double`-s if not already...but could populate the set with `BigDecimal`-s for simplicty
* add `decimalValue = convertStringToBigDecimal(value)`
* use `decimalValue` to lookup in the map

Coudln't this avoid the tree build penalty/multiple comparisions/normalizations during every comparision?

Reply:
The set is built once per query, not every comparison, however fair point about lookup speed using a plain HashSet for predicates, though there is a fair bit of complexity here that needs careful consideration and probably a bunch of measurement I was looking to avoid, at least until the bug is fixed.

`TypedInFilter` currently is constructed with a sorted `List` of match values, the sortedness is very important for computing indexes. We definitely optimized for that use case over the predicate filtering use case. 

Right now, when the column type (predicate type) matches the filters `matchValueType`, we just use the sorted list directly, and `Collections.binarySearch`, so that no set is constructed at all, though it has this same logn lookup speed that you're complaining about here, which is again a fair complaint.

As mentioned before though, we do need the sorted list, so it can't really go away, so at best we can _also_ build a hashset, so the next question is where and how to do it.

Its really not that awesome to be computing any sets in the 'makePredicate' functions because technically if several segments are in the processing pool they would be blocked while one of them builds the sets, which is kind of sad.

We could consider building the "default" set (meaning set of the matchValueType) outside of the predicate factory, as part of `TypedInFilter` constructor, though that downside is it guarantees double memory footprint instead of just when matchValueType doesn't match the column predicate type, along with the added up front cost of constructing it.

When building the set we also need to coerce the values to handle any ugly that happens during json deserialization, such as longs coming in as ints or doubles/floats or whatever, which isn't really a problem when computing indexes or with the current predicates using `Collections.binarySearch` because the match type comparator handles that (not to be confused with the `StringComparators.numeric` added in this PR, which is string predicate + number match value).

Anyway, this would let same predicate and `matchValueType` plus these string to numeric comparisons use that default set which would be computed up front, and then only have to deal with the more blocky other set construction for stranger cases of `matchValueType` not matching predicate type.

The predicate maker functions currently build sets backwards, to try to cast the match values of `matchValueType` to the predicate type, on the assumption that the number of values the predicate is going to be evaluated against is going to be a lot bigger than the matchValue set size (millions of rows vs however big the in filter is). I feel like there still seems to be some benefit to converting the set when safe rather than converting every row value (like number predicates with string matchValues), so I don't think we necessarily would _always_ want to use a hashset of the match values and convert the row values either, because it means we always have to cast every row value the predicate is processing. I also don't think we want to build all of these combinations up front, since the expense would be too much if things are never used.

Anyway, all of this is why I was suggesting that longer term it makes sense to make virtual columns more efficient and also not eat the casts in the SQL layer, so that way the predicate chosen would always match the `matchValueType`. However, still we have an open question of whether we should build a hashset up front or not, and the answer is i don't really know yet, maybe?

I guess all of this is why I was suggesting for comparison against the behavior of `"sqlUseBoundAndSelectors":true`, which makes an `OrFilter` of `BoundFilter` for `IN`, and a single `BoundFilter` for numeric equality using `StringComparators.NUMERIC`, so that this bug can be fixed without answering all of the other hard questions right now, because the current behavior is simply incorrect. However, I can at least try to explore building a same type hashset to cut down on the sets built inside the supplier and take some measurements if you view it as a blocker to fixing this bug.